packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -837,6 +860,181 @@ function rerenderReducer<S, I, A>(
+      markUnprocessedUpdateTime(pendingExpirationTime);
+    }
+  } else {
+    // If there are no pending updates, we should still check the work-in-progress version.
+    // It's possible this source (or the part we are reading from) has just not yet been subscribed to.
+    const lastReadVersion = getWorkInProgressVersion(source, isPrimaryRenderer);
+    if (lastReadVersion === null) {
+      // This is the only case where we need to actually update the version number.
+      // A changed version number for a new source will throw and restart the render,
+      // at which point the work-in-progress version Map will be reinitialized.
+      // Once a source has been subscribed to, we use expiration time to determine safety.
+      setWorkInProgressVersion(source, version, isPrimaryRenderer);
+
+      isSafeToReadFromSource = true;
+    } else {
+      isSafeToReadFromSource = lastReadVersion === version;
+    }
+  }
+
+  let prevMemoizedState = ((hook.memoizedState: any): ?MutableSourceState<
+    Source,
+    Snapshot,
+  >);
+  let snapshot = ((null: any): Snapshot);
+
+  if (isSafeToReadFromSource) {
+    snapshot = getSnapshot(source._source);
+  } else {
+    // Since we can't read safely, can we reuse a previous snapshot?
+    if (
+      prevMemoizedState != null &&
+      prevMemoizedState.config === config &&
+      prevMemoizedState.source === source
+    ) {
+      snapshot = prevMemoizedState.snapshot;
+    } else {
+      // We can't read from the source and we can't reuse the snapshot,
+      // so the only option left is to throw and restart the render.
+      // This error should cause React to restart work on this root,
+      // and flush all pending udpates (including any pending source updates).
+      // This error won't be user-visible unless we throw again during re-render,
+      // but we should not do this since the retry-render will be synchronous.
+      // It would be a React error if this message was ever user visible.
+      throw Error(
+        'Cannot read from mutable source during the current render without tearing. ' +
+          'This is a bug in React. Please file an issue.',
+      );
+    }
+  }
+
+  const prevSource =
+    prevMemoizedState != null ? prevMemoizedState.source : null;
+  const destroy =
+    prevMemoizedState != null ? prevMemoizedState.destroy : undefined;
+
+  const memoizedState: MutableSourceState<Source, Snapshot> = {
+    config,
+    destroy,
+    snapshot,
+    source,
+  };
+
+  hook.memoizedState = memoizedState;
+
+  if (prevSource !== source) {
+    const fiber = currentlyRenderingFiber;
+    const root = currentlyRenderingRoot;
+
+    const create = () => {
+      const scheduleUpdate = () => {
+        const alreadyScheduledExpirationTime = root.mutableSourcePendingUpdateMap.get(
+          source,
+        );
+
+        // If an update is already scheduled for this source, re-use the same priority.
+        if (alreadyScheduledExpirationTime !== undefined) {
```

Comment:
This check means that if you update at one priority, then again at a different priority, the first priority always wins. We should support multiple concurrent priorities, which is solved by switching to a state queue. (The test case I described in my other comment would cover this.)

Reply:
Yes, my thinking on this was that scheduling more than one update is only valid if we eagerly read snapshot values. I think the key question is whether or not that's a positive change. If it is, then going with an update queue makes sense.